test/fuzz/transformation_changing_memory_semantics_test.cpp

```diff
@@ -0,0 +1,482 @@
+
+  // Bad: The OpAtomicExchange(read-modify-write) memory semantics value cannot
+  // be changed from Acquire to Release.
+  ASSERT_FALSE(TransformationChangingMemorySemantics(
+                   MakeInstructionDescriptor(23, SpvOpAtomicExchange, 0), 0, 99)
+                   .IsApplicable(context.get(), transformation_context));
+
+  // Bad: The SpvOpAtomicCompareExchange(read-modify-write) takes two memory
+  // semantics, unequal can't acquire/release value.
+  ASSERT_FALSE(
+      TransformationChangingMemorySemantics(
+          MakeInstructionDescriptor(25, SpvOpAtomicCompareExchange, 0), 1, 100)
+          .IsApplicable(context.get(), transformation_context));
+}
+
+TEST(TransformationChangingMemorySemanticsTest, AtomicInstructionsTestCases) {
+  const std::string shader = R"(
+               OpCapability Shader
+               OpCapability Int8
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main"
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource ESSL 320
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %6 = OpTypeInt 32 1
+          %7 = OpTypeInt 8 1
+          %9 = OpTypeInt 32 0
+          %8 = OpTypeStruct %6
+         %10 = OpTypePointer StorageBuffer %8
+         %11 = OpVariable %10 StorageBuffer
+         %18 = OpConstant %9 1
+         %12 = OpConstant %6 0
+         %13 = OpTypePointer StorageBuffer %6
+         %15 = OpConstant %6 4
+         %16 = OpConstant %6 7
+         %17 = OpConstant %7 4
+         %250 = OpConstant %6 66  ; Acquire | UniformMemory
+         %26 = OpConstant %6 258 ; Acquire | WorkgroupMemory
+         %27 = OpConstant %6 68  ; Release | UniformMemory
+         %28 = OpConstant %6 80  ; SequentiallyConsistent | UniformMemory
+         %29 = OpConstant %6 64  ; None | UniformMemory
+         %30 = OpConstant %6 256 ; None | WorkgroupMemory
+         %31 = OpConstant %6 72 ; AcquireRelease | UniformMemory
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+         %14 = OpAccessChain %13 %11 %12
+         %21 = OpAtomicLoad %6 %14 %15 %250
+         %22 = OpAtomicLoad %6 %14 %15 %29
+         %23 = OpAtomicLoad %6 %14 %15 %27
+         %32 = OpAtomicExchange %6 %14 %15 %31 %16
+         %24 = OpAccessChain %13 %11 %12
+               OpAtomicStore %14 %15 %27 %16
+               OpReturn
+               OpFunctionEnd
+  )";
+  const auto env = SPV_ENV_UNIVERSAL_1_3;
+  const auto consumer = nullptr;
+  const auto context = BuildModule(env, consumer, shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+  // Bad: Memory semantics can't change with the same value.
```

Comment:
These look like good tests. But I think it would be clearer if you gave the full details in the comment. E.g.
// Bad transformation:
// OpAtomicLoad Acquire | UniformMemory
// to:
// OpAtomicLoad Acquire | WorkgroupMemory

Also, in hindsight, I should have suggested two different tests:
// Bad transformation: (higher bits must not change)
// OpAtomicLoad Acquire | UniformMemory
// to:
// OpAtomicLoad SequentiallyConsistent | WorkgroupMemory

// Bad transformation: (lower bits must change)
// OpAtomicLoad Acquire | UniformMemory
// to:
// OpAtomicLoad Acquire | UniformMemory

But this was my mistake, so no need to do this right now.

Reply:
Done.